### PR TITLE
Add caching dupe change

### LINKED ISSUE #340: [CI] Deduplicate `actions/cache` directives
We have a `generate_cache` job in CI which pre-populates a cache that is re-used by other jobs:

https://github.com/google/zerocopy/blob/3bb9a54ff458e104f7eb24beda7f8fb09794e1a5/.github/workflows/ci.yml#L410-L442

In each dependent job, we do the following:
- Use `needs: generate_cache`
- Use the following `actions/cache` action configuration:
  https://github.com/google/zerocopy/blob/3bb9a54ff458e104f7eb24beda7f8fb09794e1a5/.github/workflows/ci.yml#L77-L81

As of this writing, we copy that exact text five times. It'd be great if we could reduce the duplication.

## Mentoring instructions

As [proposed here](https://github.com/google/zerocopy/issues/340#issuecomment-1747242751) by @mgeisler:
- [ ] Create a separate `cache.yml` file in `.github/workflows`
- [ ] Populate it with a job that performs these steps:
  
  https://github.com/google/zerocopy/blob/3bb9a54ff458e104f7eb24beda7f8fb09794e1a5/.github/workflows/ci.yml#L77-L81
- [ ] Replace uses of that pattern in `ci.yml` with an invocation of that file, ie:
  ```yaml
  - name: Populate cache
    uses: `.github/workflows/cache.yml`
  ```

<details>
<summary>Old/obsolete mentoring instructions</summary>

As discussed in [this StackOverflow comment](https://stackoverflow.com/questions/77053736/give-alias-to-github-actions-uses-step?noredirect=1#comment135837054_77053736), 

> You can extract your own actions: [docs.github.com/en/actions/creating-actions](https://docs.github.com/en/actions/creating-actions). YAML's own support for this isn't available in GHA though: [github.com/actions/runner/issues/1182](https://github.com/actions/runner/issues/1182).

The task is to figure out whether there's a way to use that technique to define a new action in `ci.yml` that we can re-use in other steps. If there's an easy, simple syntax for this, we should do it. If it would require a lot of complexity, then it's not worth it.

</details>

### PR DESCRIPTION
**Pull request by dorryspears**
This closes #340.  I tried putting the yaml in the workflow folder for the new cache job but it wasn't able to find it, so I made it an action.

### CODE REVIEW

**Comment by joshlf** [comment]
Thanks for this, @dorryspears !

I'm not a GitHub Actions expert, so I may be off the mark here, but will making this a "step" rather than a `needs` dependency mean that this action is executed multiple times? Currently (on `main`), it executes once at the beginning of the job run and then the results are re-used by all steps, which helps things run faster.

**Comment by dorryspears** [comment]
@joshlf You are right.  I now understand what the CI was doing.  I rearranged my code and added back the generate cache step.  That action just replaces the 

```yaml
- uses: actions/cache@13aacd865c20de90d75de3b17ebe84f7a17d57d2 # v4.0.0
      with:
        path: |
          ~/.cargo/
        key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.toml') }}
``` 
step in the existing jobs.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -519,7 +504,7 @@ jobs:
           cargo install cargo-readme --version 3.2.0 &> /dev/null || true
           cargo install --locked kani-verifier       &> /dev/null || true
           cargo kani setup                           &> /dev/null || true
-
+
```
Comment (joshlf) [inline .github/workflows/ci.yml:507]:
```suggestion

```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -544,7 +529,7 @@ jobs:
             <(cargo metadata --format-version 1 | \
               jq -r ".packages[] | select(.name == \"zerocopy\").metadata.\"build-rs\" | keys | .[]" | \
               sort -u)
-
+
```
Comment (joshlf) [inline .github/workflows/ci.yml:532]:
```suggestion

```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -485,20 +473,17 @@ jobs:
           assert-match "$zerocopy_derive_dep_ver" "$zerocopy_derive_targeted_ver" \
             "Same crate version ($zerocopy_derive_dep_ver) found for optional and targeted zerocopy-derive dependency." \
             "Different crate versions found for optional ($zerocopy_derive_dep_ver) and targeted ($zerocopy_derive_targeted_ver) dependency."
-
+
```
Comment (joshlf) [inline .github/workflows/ci.yml:476]:
```suggestion

```

**Review by joshlf — CHANGES_REQUESTED** [review]
Okay awesome, thanks! A few nits regarding stray whitespace, but this looks good!

**Review by joshlf — APPROVED** [review]

**Pull request merged**